src/backend/optimizer/plan/createplan.c

```diff
@@ -6688,10 +6688,13 @@ adjust_modifytable_flow(PlannerInfo *root, ModifyTable *node, List *is_split_upd
 				 */
 				if (optimizer_replicated_table_insert &&
 					subplan->flow->flotype == FLOW_SINGLETON &&
-					subplan->flow->locustype == CdbLocusType_SegmentGeneral &&
-					!contain_volatile_functions((Node *)subplan->targetlist))
+					subplan->flow->locustype == CdbLocusType_SegmentGeneral)
 				{
-					if (subplan->flow->numsegments >= targetPolicy->numsegments)
+					if (contain_volatile_functions((Node *)subplan->targetlist))
```

Comment:
Why do you still check only target list here? What about grouping and having?

Reply:
havingQual added

Reply:
Currently, you check target list and quals for this case. Please, explain why should we not check having clause quals?

Reply:
having added but could not find example yet

Reply:
All that you need is a query with some node above Aggregation. There are dozens of examples in the regression test output files. For example, you may use aggregate with ordering that produce Sort node above.

Reply:
I mean that broadcast is made in such case and without patch:
```sql
explain (costs off, verbose) insert into t_replicate_dst (i) select i from t_replicate_src group by i having random() > 0;
                                       QUERY PLAN                                       
----------------------------------------------------------------------------------------
 Insert on public.t_replicate_dst
   ->  Broadcast Motion 1:3  (slice1; segments: 1)
         Output: ((nextval('t_replicate_dst_id_seq'::regclass))::integer), "*SELECT*".i
         ->  Subquery Scan on "*SELECT*"
               Output: nextval('t_replicate_dst_id_seq'::regclass), "*SELECT*".i
               ->  HashAggregate
                     Output: t_replicate_src.i
                     Group Key: t_replicate_src.i
                     Filter: (random() > '0'::double precision)
                     ->  Seq Scan on public.t_replicate_src
                           Output: t_replicate_src.i
 Optimizer: Postgres query optimizer
(12 rows)
```

Reply:
Of course, because having was checked in the `set_subquery_path`. This part of code handles cases when subqueries are removed as trivial. Is it possible to have trivial subplan with aggregation?

Reply:
if it is removed, from when there is broadcast?

Reply:
having removed because it s catched by set_subquery_pathlist, because it is not trivial